base/sort.jl

```diff
@@ -1298,6 +1298,53 @@ function sort!(A::AbstractArray;
     end
     A
 end
+"""
+    sortperm(A; dims::Integer, alg::Algorithm=DEFAULT_UNSTABLE, lt=isless, by=identity, rev::Bool=false, order::Ordering=Forward)
+
+    Return a permutation vector `I` that puts `v[I]` in sorted order along the given dimension. The order is specified
+    using the same keywords as [`sort!`](@ref). The permutation is guaranteed to be stable even
+    if the sorting algorithm is unstable, meaning that indices of equal elements appear in
+    ascending order.
+
+    To sort slices of an array, refer to [`sortslices`](@ref).
+
+    # Examples
+    ```jldoctest
+    julia> A = [4 3; 1 2]
+    2×2 Matrix{Int64}:
+     4  3
+     1  2
+    
+    julia> sortperm(A, dims = 1)
+    2×2 Matrix{Int64}:
+     2  4
+     1  3
+    
+    julia> sortperm(A, dims = 2)
+    2×2 Matrix{Int64}:
+     3  1
+     2  4
+```
+"""
+function sortperm(A::AbstractArray; dims::Integer,
+    alg::Algorithm=DEFAULT_UNSTABLE,
+    lt=isless,
+    by=identity,
+    rev::Union{Bool,Nothing}=nothing,
+    order::Ordering=Forward)
+    P = mapslices(x -> sortperm(x; rev, lt, alg, by, order), A, dims=dims)
```

Comment:
Is it possible to sort the indices that correspond to `A[...,:,...]` with `order=Perm(ord(lt, by, rev, order), A)` to make the following 12 lines unnecessary? 

Maybe `OffsetArray`s or `view`s would be useful for this, but they may not be necessary.

Reply:
And if it's not possible to elegantly and efficiently get the final indices directly from a call to sort/sortperm, it might also be possible to correct them with a simple multiplication and broadcasted addition.

Reply:
Thank you for the suggestion!

I have come up with a couple ideas. 

The function `sortperm_new_2` (below) does it in 3 lines as you suggested, it is slightly faster than the original as well.

However, if we loop over the slices ourselves, it's about 10x faster (at least on 10x10x10 matrices), as shown by `sortperm_2` below. I have not yet looked into why `mapslices` performs so poorly here, but I figure the problem is in there somewhere.


<details>
<summary>Full test script</summary>

```
function sortperm_original(A::AbstractArray; dims::Integer, rev::Bool=false)
    P = mapslices(x -> sortperm(x; rev=rev), A; dims=dims)
    Rpre = CartesianIndices(size(A)[1:dims-1])
    Rpost = CartesianIndices(size(A)[dims+1:end])
    _sortperm!(P, Rpre, dims, Rpost)
    return P
end

#Helper function to introduce function barrier on CartesianIndices
@inline function _sortperm!(P::AbstractArray, Rpre, dims, Rpost)
    for Ipost in Rpost, i = axes(P, dims), Ipre in Rpre
        P[Ipre, i, Ipost] = LinearIndices(P)[Ipre, P[Ipre, i, Ipost], Ipost]
    end
end

function sortperm_new(A::AbstractArray; dims::Integer)
    Rpre = CartesianIndices(size(A)[1:dims-1])
    Rpost = CartesianIndices(size(A)[dims+1:end])
    ix = similar(A)
    ix .= LinearIndices(A)
    ord = Base.Perm(
        Base.ord(Base.isless, identity, nothing, Base.Order.Forward),
        vec(reshape(A,(:,1)))
    )
    _sortperm_new!(ix, A, Rpre, dims, Rpost,ord)
    return ix
end

@inline function _sortperm_new!(ix::AbstractArray,A::AbstractArray, Rpre, dims, Rpost, order)
    for Ipost in Rpost, Ipre in Rpre
        sort!(view(ix,Ipre, :, Ipost);order)
    end
end

function sortperm_new_2(A::AbstractArray; dims::Integer)
    ord = Base.ord(Base.isless, identity, nothing, Base.Order.Forward)
    order = Base.Perm(ord, vec(reshape(A, (1,:))))
    return mapslices(x -> sort(x; order), LinearIndices(A); dims=dims)
end

using BenchmarkTools
function test()
    for dims in 1:3
        a = rand(1:10,10,10,10)
        @info "dims = $dims"
        @btime sortperm_original($a; dims=$dims)
        @btime sortperm_new($a; dims=$dims)
        @btime sortperm_new_2($a; dims=$dims)

        perm_1 = sortperm_original(a; dims=dims)
        perm_2 = sortperm_new(a; dims=dims)
        perm_3 = sortperm_new_2(a; dims=dims)
        @assert perm_1 == perm_2 == perm_3
    end
end

test()
```

</details>

Reply:
Do you think we should use the faster option? or the more concise one, and I'll see if I can make an issue/PR for `mapslices`?

Reply:
Timing with #40996 doubles the performance of `sortperm_new_2`, but `mapslices` is still type unstable. Since the longer `sortperm_2` is type stable and much more performant, I will make a commit with that one that also addresses your other comments.

Reply:
I've been referring to the existing `sort!(A::AbstractArray; dims::Integer, ...)` as a model, but it is not an infallible model. Your work (`sortperm_new`) already has less dynamic dispatch. I still find it helpful to reference that existing code, though.

Reply:
>Ideally, both. The performance issue with mapslices comes from the fact that dims is not known at compile time. We can put the dynamic dispatch (both order and dims) at the very top of the function with:

Hmm I do not see much improvement in performance in the `mapslices` version when putting the dynamic dispatch into another function as you are doing. See below for my test script. `descend`ing with Cthulhu confirms that the mapslices call in `sortperm_val` is still not being concretely typed, and #40996 does not fix the issue. Did I miss something?

My results for the script below are:

```julia
[ Info: dims = 1
[ Info: original
  50.050 μs (736 allocations: 39.39 KiB)
[ Info: PR
  6.336 μs (8 allocations: 8.19 KiB)
[ Info: shorter version w/ mapslices
  38.470 μs (739 allocations: 42.66 KiB)
[ Info: shorter w/ mapslices and val barrier
  39.290 μs (739 allocations: 42.66 KiB)
```

Ah it's actually more like a 5 fold speedup, it is 10-fold on very small matrices (4x4x4), my mistake.


>You're welcome to include something like that in this PR, or call it out of scope and you or I can make the improvement in a different PR. This one can certainly merge without the 10x speedup in sort!(A, dims) for small Arrays.

I think it's best to do that in another PR. Let me know if you have any other comments on this PR as-is or whether it is good to merge!


<details>

<summary> benchmark script </summary>

```julia
function sortperm_original(A::AbstractArray; dims::Integer, rev::Bool=false)
    P = mapslices(x -> sortperm(x; rev=rev), A; dims=dims)
    Rpre = CartesianIndices(size(A)[1:dims-1])
    Rpost = CartesianIndices(size(A)[dims+1:end])
    _sortperm!(P, Rpre, dims, Rpost)
    return P
end

#Helper function to introduce function barrier on CartesianIndices
@inline function _sortperm!(P::AbstractArray, Rpre, dims, Rpost)
    for Ipost in Rpost, i = axes(P, dims), Ipre in Rpre
        P[Ipre, i, Ipost] = LinearIndices(P)[Ipre, P[Ipre, i, Ipost], Ipost]
    end
end

function sortperm_new(A::AbstractArray; dims::Integer)
    Rpre = CartesianIndices(size(A)[1:dims-1])
    Rpost = CartesianIndices(size(A)[dims+1:end])
    ix = similar(Array{Int}, axes(A))
    ix .= LinearIndices(A)
    ord = Base.Perm(
        Base.ord(Base.isless, identity, nothing, Base.Order.Forward),
        vec(reshape(A, (:, 1)))
    )
    _sortperm_new!(ix, A, Rpre, dims, Rpost, ord)
    return ix
end

@inline function _sortperm_new!(ix::AbstractArray, A::AbstractArray, Rpre, dims, Rpost, order)
    for Ipost in Rpost, Ipre in Rpre
        sort!(view(ix, Ipre, :, Ipost); order)
    end
end

function sortperm_new_2(A::AbstractArray{T,N}; dims::Integer) where {T,N}
    _sortperm_new_2(A, Val(dims))
end

function _sortperm_new_2(A::AbstractArray{T,N}, ::Val{K}) where {T,N,K}
    ord = Base.ord(Base.isless, identity, nothing, Base.Order.Forward)
    order = Base.Perm(ord, vec(reshape(A, (1, :))))
    P::Array{Int,N} = mapslices(x -> sort(x; order), LinearIndices(A); dims=K)
    return P
end

function sortperm_val(A::AbstractArray{T,N}; dims::Integer) where {T,N}
    _sortperm_val(A, Val(dims))
end

function _sortperm_val(A::AbstractArray{T,N}, ::Val{K}) where {T,N,K}
    ord = Base.ord(Base.isless, identity, nothing, Base.Order.Forward)
    order = Base.Perm(ord, vec(reshape(A, (1, :))))
    P::Array{Int,N} = mapslices(x -> sort(x; order), LinearIndices(A); dims=K)
    return P
end
using BenchmarkTools
function test()
    for dims in 1:3
        a = rand(10, 10, 10)
        @info "dims = $dims"
        @info "original"
        @btime sortperm_original($a; dims=$dims)
        @info "PR"
        @btime sortperm_new($a; dims=$dims)
        @info "shorter version w/ mapslices"
        @btime sortperm_new_2($a; dims=$dims)
        @info "shorter w/ mapslices and val barrier"
        @btime sortperm_val($a; dims=$dims)

        perm_1 = sortperm_original(a; dims=dims)
        perm_2 = sortperm_new(a; dims=dims)
        perm_3 = sortperm_new_2(a; dims=dims)
        perm_4 = sortperm_val(a; dims=dims)
        @assert perm_1 == perm_2 == perm_3 == perm_4
    end
end

test()
</details>

Reply:
> Hmm I do not see much improvement in performance in the `mapslices` version when putting the dynamic dispatch into another function as you are doing.

Oops! You're right. I noticed a big improvement in `sort!` but `sort!` is already using a loop rather than `mapslices`. We could add a todo here to switch to `mapslices` eventually if/when it becomes as fast/faster than a loop, but until then, I think you are right to simply not use `mapslices` at all.

Reply:
Thank you for your thoroughness and your benchmarks!